File: examples/csharp-dotnet/request/request.cs
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/**
 * @file request.cs
 *
 * This file contains an example demonstrating how to issue a request message
 * and handle a coinciding reply message. This example program is intended to
 * be run after starting up either the 'reply' or 'reply_async' example program.
 */

using GMSEC.API;
using GMSEC.API.UTIL;
using GMSEC.API.MIST;
using System;

class request
{
	static private String DEFAULT_REQUEST_SUBJECT = "GMSEC.TEST.REQUEST";

	static int Main(string[] args)
	{
		if (args.Length < 1)
		{
			Console.WriteLine("usage request.exe mw-id=<middleware ID>");
			return -1;
		}

		//o Load the command-line input into a GMSEC Config object
		// A Config object is basically a key-value pair map which is used to
		// pass configuration options into objects such as Connections,
		// ConnectionManagers, Subscribe and Publish function calls, Messages,
		// etc.
		Config config = new Config(args);

		//o Since this example program uses an invalid message, we ensure the
		//  validation check is disabled.
		config.AddValue("gmsec-msg-content-validate-all", "false");

		//o If it was not specified in the command-line arguments, set LOGLEVEL
		// to 'INFO' and LOGFILE to 'stdout' to allow the program report output
		// on the terminal/command line
		InitializeLogging(config);

		//o Print the GMSEC API version number using the GMSEC Logging
		// interface
		// This is useful for determining which version of the API is
		// configured within the environment
		Log.Info(ConnectionManager.GetAPIVersion());

		try
		{
			//o Create the ConnectionManager
			ConnectionManager connMgr = new ConnectionManager(config);

			//o Open the connection to the middleware
			Log.Info("Opening the connection to the middleware server");
			connMgr.Initialize();

			//o Output middleware client library version
			Log.Info(connMgr.GetLibraryVersion());

			//o Output information
			Log.Info("Issuing a request using the DEFAULT_REQUEST_SUBJECT '" + DEFAULT_REQUEST_SUBJECT);

			//o Create message
			using(Message requestMsg = new Message(DEFAULT_REQUEST_SUBJECT, Message.MessageKind.REQUEST))
			{

				//o Add fields to message
				requestMsg.AddField(new StringField("QUESTION", "Is there anyone out there?"));
				requestMsg.AddField(new StringField("COMPONENT", "request"));

				//o Display XML representation of request message
				Log.Info("Sending request message:\n" + requestMsg.ToXML());

				//o Send Request Message
				// Timeout periods:
				// -1 - Wait forever
				//  0 - Return immediately
				// >0 - Time in milliseconds before timing out
				Message replyMsg = connMgr.Request(requestMsg, -1, -1);

				// Example error handling for calling request() with a timeout
				if (replyMsg != null)
				{
					// Display the XML string representation of the reply
					Log.Info("Received replyMsg:\n" + replyMsg.ToXML());

					//o Destroy the replyMsg message
					connMgr.Release(ref replyMsg);
				}
			}

			//o Disconnect from the middleware and clean up the Connection
			connMgr.Cleanup();
		}
		catch (Exception e)
		{
			Log.Error(e.ToString());
			return -1;
		}

		return 0;
	}

	static void InitializeLogging(Config config)
	{
		String logLevel = config.GetValue("LOGLEVEL");
		String logFile = config.GetValue("LOGFILE");

		if (logLevel == null)
		{
			config.AddValue("LOGLEVEL", "INFO");
		}
		if (logFile == null)
		{
			config.AddValue("LOGFILE", "STDERR");
		}
	}
}

File: examples/java/request.java
/*
 * Copyright 2007-2023 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file request.java
 *
 * @brief This example program provides a basic demonstration of issuing
 * a synchronous Request Directive message to the GMSEC Bus, and then
 * handling the receipt of a Response Directive message.
 */

import gov.nasa.gsfc.gmsec.api5.*;
import gov.nasa.gsfc.gmsec.api5.field.*;
import gov.nasa.gsfc.gmsec.api5.util.Log;

import java.util.ArrayList;
import java.util.List;


public class request
{
	public static void main(String[] args)
	{
		// Set up connection configuration options, some from the command line,
		// others hard-coded.
		Config config = new Config(args);

		// Initialize log level for output
		utility.initializeLogLevel(config);

		// Validate messages before being sent to and after being received from the bus
		config.addValue("gmsec-msg-content-validate-all", "true");

		try
		{
			// Create connection instance.
			Connection conn = Connection.create(config);

			// Set up standard fields within the MessageFactory associated
			// with the connection object.
			setStandardFields(conn.getMessageFactory(), "REQUEST");

			// Establish connection to the GMSEC Bus.
			conn.connect();

			// Output information regarding the API version and connection
			Log.info(Connection.getAPIVersion());
			Log.info("Middleware version = " + conn.getLibraryVersion());

			// Create a Request Directive message using the MessageFactory.
			// Set both required and optional fields within the message.
			Message reqMsg = conn.getMessageFactory().createMessage("REQ.DIR");
			reqMsg.addField(new StringField("DESTINATION-COMPONENT", "REPLY", true));   // to whom is the request intended for
			reqMsg.addField("DIRECTIVE-STRING", "Turn on the lights");                  // operation to perform
			reqMsg.addField("RESPONSE", true);                                          // a response message is requested
			reqMsg.addField("REQUEST-ID", new U16(0));

			// Output in XML what we intend to send
			Log.info("Sending Request Message:\n" + reqMsg.toXML());

			// Timeout and republish periods have been arbitrarily chosen to be 5 seconds.
			// The republish period could also be set to GMSEC_REQUEST_REPUBLISH_NEVER to
			// prevent re-sending request message should a timeout occur.
			int timeout   = 5000;
			int republish = 5000;

			// Issue Request Directive.
			// Note, this is a blocking call (i.e. a synchronous operation).
			Message rspMsg = conn.request(reqMsg, timeout, republish);

			// We have returned from the previous blocking call; check if we have a message.
			if (null != rspMsg)
			{
				Log.info("Received Response Directive message:\n" + rspMsg.toXML());

				// Acknowledge processing the response message
				rspMsg.acknowledge();

				// Destroy the response message
				Message.destroy(rspMsg);
			}
			else
			{
				// This else-block is reached if a response message is not received.
				// If the republish period is set to a time period other than
				// GMSEC_REQUEST_REPUBLISH_NEVER, then this block will never be reached.
				Log.warning("Time out; no Response Directive message received");
			}

			// Disconnect from the GMSEC Bus, and terminate subscriptions.
			conn.disconnect();

			// Destroy the connection instance
			Connection.destroy(conn);
		}
		catch (GmsecException e)
		{
			// If here, the most likely case is the middleware could not be
			// deduced, or a failure occurred when attempting to connect to
			// the GMSEC Bus.
			Log.error("GmsecException: " + e.getMessage());
		}
	}


	// Convenience function for associating standard (i.e. common) fields
	// with the MessageFactory which will be included automatically within
	// all messages created using the factory.
	private static void setStandardFields(MessageFactory factory, String component)
	{
		StringField field1 = new StringField("MISSION-ID", "MY-MISSION", true);
		StringField field2 = new StringField("SAT-ID-PHYSICAL", "MY-SAT-ID", true);
		StringField field3 = new StringField("SAT-ID-LOGICAL", "MY-SAT-ID", true);
		StringField field4 = new StringField("CONSTELLATION-ID", "MY-CNST", true);
		StringField field5 = new StringField("FACILITY", "MY-FACILITY", true);
		StringField field6 = new StringField("DOMAIN1", "MY-DOMAIN-1", true);
		StringField field7 = new StringField("DOMAIN2", "MY-DOMAIN-2", true);
		StringField field8 = new StringField("COMPONENT", component, true);

		List<Field> standardFields = new ArrayList<Field>();
		standardFields.add(field1);
		standardFields.add(field2);
		standardFields.add(field3);
		standardFields.add(field4);
		standardFields.add(field5);
		standardFields.add(field6);
		standardFields.add(field7);
		standardFields.add(field8);

		factory.setStandardFields(standardFields);
	}
}
